### ISSUE TITLE
Config sometimes not updated after restart of the manage-server

### ISSUE THREAD

**Issue by luckyswede** [googleforgames/quilkin#682]
I still sometimes get the same behavior as in #660, where the proxy stops getting updates about new gameserver instances after a restart of the xds mgmt server.

I dont have any detailed logs atm, but I'll try to provide some.

Quilkin version: us-docker.pkg.dev/quilkin/ci/quilkin:0.5.0-dev-38a16e2

**Comment by markmandel** [comment]
Doh! I had hoped I had fixed it.

Logs, particularly from the proxy would definitely help.

You also mentioned "after restart" -- what is causing the xDS provider to restart? Is that you restarting it, or is it failing somehow?

**Comment by XAMPPRocky** [comment]
I've also been seeing this, I'll investigate further on Monday, I think the retry logic in the proxy isn't retrying.

**Comment by luckyswede** [comment]
> You also mentioned "after restart" -- what is causing the xDS provider to restart? Is that you restarting it, or is it failing somehow?

Yeah, this time it was caused by a container restart of the xDS provider. Not sure why it restarted, I havent had time to look into it.

**Comment by XAMPPRocky** [comment]
> This is the last log from the xds server before it restarted

That makes sense why it would cause the management server to restart, the bigger issue IMO, is that after the restart, the proxy stops getting info after the restart. The proxy should be resilient to losing the management server.

---

I can't replicate this locally just from killing and restarting the management server, multiple connected proxies just work. I'm going to try and see if it's maybe specific to linux or k8s.

**Comment by luckyswede** [comment]
I'm able to reproduce it in GKE by simply deleting the mgmt server.
It's running behind a ClusterIP service in my setup.

**Comment by XAMPPRocky** [comment]
@luckyswede I can't replicate it by just deleting it in GCP. Can you set `RUST_LOG=info,quilkin=trace`, and provide the full log for both services?

**Comment by luckyswede** [comment]
@XAMPPRocky here is an example. Look for token `aENESFlnTg==`
[proxy.log](https://github.com/googleforgames/quilkin/files/10428680/proxy.log)
[xds.log](https://github.com/googleforgames/quilkin/files/10428681/xds.log)

**Comment by XAMPPRocky** [comment]
@luckyswede Thank you so much, these are very helpful.

The proxy not retrying seemed to be a red herring, it seems to be the xds server not providing an up to date response. I'll investigate more tomorrow.

**Comment by XAMPPRocky** [comment]
Okay I think I've found what the issue is, but I haven't figured it out how to fix it and need to implement some other parts which are higher priority for us right now, so I've added a hot fix that should ensure they don't drift apart for too long, and added a warning that should help diagnose the root of the problem.

The issue seems to be here in the rx path of the ids server stream. From the logs above it seems that sometimes, `rx.changed()` is not being fired when it should, `Slot<T>` is pushing the update, but it seems like it's not then being turnt into a xDS response.

@luckyswede Would you be able to do the same thing with the image made from #684? That should give us more of indication of where the update is failing to be triggered.

https://github.com/googleforgames/quilkin/blob/72176a191ffc3a597e3834ee1d0090b30caf81d4/src/xds/server.rs#L168-L187

**Comment by luckyswede** [comment]
@XAMPPRocky Yes for sure, nice that you found the problem! What log entry should I look for?

**Comment by XAMPPRocky** [comment]
@luckyswede You should look for the same thing as last time, a proxy not having an update so it's missing a token and dropping packets, and you see that token being added in the management server.

**Comment by luckyswede** [comment]
@XAMPPRocky here u go, logs when the same thing happens with build from 684
[proxy.log](https://github.com/googleforgames/quilkin/files/10449842/proxy.log)
[xds.log](https://github.com/googleforgames/quilkin/files/10449843/xds.log)

**Comment by XAMPPRocky** [comment]
Well this does eliminate one possibility (which is that the change requests notifications were being sliently dropped). It seems like `rx.changed()` is not enough, maybe we need to track individual changes instead of just checking if it's changed since we last looked.
